### PR TITLE
Change sqlite query-result to a pseudo resource

### PR DESCRIPTION
**Pull request by rylev**
Hot on the heels of #1408 landing, this PR proposes a change to the API that makes `query-result` (i.e., the return type from executing a sqlite statement) into a pseudo resource. This changes the API so that instead of receiving all of the results at once, the user must instead either collect the results through some iteration or query from a specific index of the `query-result`.

## What this looks like in code

On the Rust side of things, this is largely hidden behind an iterator:

```rust
   let todos = conn
         .execute(&format!("SELECT * FROM todos;"), &[])?
         .rows()
         // the only change from previous version is the need for the `?`
         // since each iteration is technically failable (more on this below)
         .map(|r| -> anyhow::Result<Todo> { r?.try_into() }) 
         .collect::<anyhow::Result<Vec<Todo>>>()?;
```

## Downsides

There are a few known downsides to this approach:

* This is ever so slightly more complicated as the user is no longer given a simple data structure
  * As a consequence of this, the SDKs need to do more work to hide some of this complexity.
  * This will be reduced in the future when resources, streams, etc are introduced
* Each call to get the next value in `query-result` can fail since we technically never know on the host side if they handle we're being given is still valid. 
  * I believe this issue would go away with resources.

## Upsides

* The guest has much less risk of running into memory issues if the `query-result` set is very large. 
* This is more future proof compatible with resources and streams

### CODE REVIEW

File: docs/content/sips/013-sqlite.md [hunk 1]
```diff
@@ -74,13 +76,17 @@ query: func(conn: connection, query: string, parameters: list<value>) -> expecte
...
-  rows: list<row-result>,
-}
+// A result of making a query
+type query-result = u32
+
+// Get columns for a given `query-result`
+get-columns: func(query-result: query-result) -> expected<list<string>, error>
+
+// Get a `row-result` at a given index for a given `query-result`
+get-row-result: func(query-result: query-result, index: u32) -> expected<option<row-result>, error>
```
Comment (lann) [inline docs/content/sips/013-sqlite.md:86]: Do sqlite results support random access? I thought they had cursor/iterator semantics?

Comment (rylev) [reply docs/content/sips/013-sqlite.md:86]: The implementation does not mirror the underlying semantics of sqlite - we just keep things in memory on the host side which makes random access possible. We could do more closely mirror the sqlite API if we want though...

File: docs/content/sips/013-sqlite.md [hunk 2]
```diff
@@ -52,6 +52,8 @@ variant error {
   access-denied,
   // The provided connection is not valid
   invalid-connection,
+  // The provided query result is not valid
+  invalid-query-result,
```
[the remark this answers is not in the document]
Comment (lann) [reply docs/content/sips/013-sqlite.md:56]: Do these even need to be errors? For pseudo-resources it seems reasonable to just trap if a handle is invalid.

File: docs/content/sips/013-sqlite.md (hunk 1 above)
Comment (itowlson) [reply docs/content/sips/013-sqlite.md:86]: +1 for cursor/iterator if we go down the one row at a time route.

File: docs/content/sips/013-sqlite.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (rylev) [reply docs/content/sips/013-sqlite.md:56]: We could also do that. Not sure what the best practice is, but it does seem reasonable to trap given that passing an invalid handle is 100% guaranteed to be programmer error.

[the remark this answers is not in the document]
Comment (lann) [reply docs/content/sips/013-sqlite.md:56]: Looking at some of the preview2 APIs I believe that is the approach they are using.

**Comment by rylev** [comment]
Going to close this as we've decided this is the right change for now. In the future we'll want to have some sort of cursor based API, but given the current limitations of the component model and the needs of users the added complexity isn't yet worth it. We'll revisit this in the future.

**Pull request closed without merging**